### PR TITLE
Revise type legalization so it can handle constant buffers

### PR DESCRIPTION
**Pull request by tfoleyNV**
The existing legalization approach with "tuples" can handle scalarizing a `struct` type with resource-type fields in it, but it had several big gaps. The most notable is that given a type that mixes uniform and resource fields, we can't just blindly scalarize things:

```
struct P {
    float4 a;
    float4 b;
    Texture2D t;
};

cbuffer C
{
    P gParam[8];
};
```

The existing code was completely ignoring the declaration of `gParam` inside `C`, but even if we fixed that issue, we'd get something like:

```
cbuffer C
{
    float4 gParam_a[8];
    float4 gParam_b[8];
};
Texture2D gParam_t[8];
```

In this case we've completely changed the layout of the uniform buffer, by switching from AOS to SOA.
Even if we could get the type layout logic and the IR to agree on this, it would be a surprise to users, and "principle of least surprise" should be a big deal on a project with as many moving parts as ours.

The right thing to do is to have legalization create a "stripped" version of the original type `P` and use that:

```
struct P_stripped {
    float4 a;
    float4 b;
};
cbuffer C
{
    P_stripped gParam[8];
};
Texture2D gParam_t[8];
```

Then at a call site, this:

```
foo(gParam);
```

becomes:

```
foo(gParam, gParam_t);
```

This is exactly how the current AST-to-AST legalization handles mixed uniform and resource types, but the way it does it involves some annoying kludges:

- That pass has a notion of a "tuple" similar to our legalization, but every tuple has an optional "primary" entry for all the uniform data, plus tuple elements for the resources, and a given field may be represented on one side, the other, or both. It makes the code for handling tuples very messy.

- That pass does the "stripping" of types by actually marking up the AST declarations (this is okay because it is constructing a new AST as it goes), so that when they get emitted certain fields don't actually show up. That is, we fix the problem with type `P` by actually *modifying* the user's declaration of `P`. That seems out of bounds for the IR.

This change fixes the problem in our IR type legalization while trying to avoid the problems of the AST-to-AST pass by using two new ideas:

1. We add a new case for `LegalType` (and `LegalVal`) that is a "pair" type, where a pair consists of both an "ordinary" type (for uniform data) and a "special" type (for resource data). E.g., after legalization, the type for `C` (which can be over-simplified to `ConstantBuffer<P>` for our purposes), will be a `LegalType::pair` where the ordinary side is `ConstantBuffer<P_stripped>` and the special side is a tuple containing the `Texture2D` field.

2. We add a new (and annoyingly hacky) AST-level type called `FilteredTupleType` which is semantically a sort of tuple type (it holds a list of elements, and the elements have their own types), but which remembers an "original type" that it was created from, and for each element remembers the field of the original type that it corresponds to. This is used to construct a type like `P_stripped` as an actual AST-level structural type.

The core logic for legalizing an aggregate type had to get more complicated just because of the new pair case, so there is now a `TupleTypeBuilder` that asists with taking an aggregate type, processing its fields, and then picking the right `LegalType` representation for the result.

Other smaller changes:

- Made the legalization logic actually legalize `PtrType<T>`. E.g., if `T` legalizes to a tuple, we need to construct a tuple of pointer types. The same exact thing needs to be applied to arrays, and any other generic type that should "distribute over" pairs/tuples.

- Made the legalization logic actually legalize `ConstantBuffer<T>` and similar. The basic idea there is if `T` maps to a pair, we wrap `ConstantBuffer<...>` around the ordinary side, and `implicitDeref` around the special side.

- Removed a bunch of `#ifdef`ed-out code from the end of `ir-legalize-types.cpp`. That was code from my first attempt at legalization that failed miserably (trying to do it via local changes and a work list instead of a global rewrite pass), but it had some code I wanted to reference when writing the version that actually got checked in (should have deleted the code earlier, though).

- Added a bunch of cases for `LegalType::none` (and the `LegalVal` equivalent) that helped simplify the logic fo the `pair` case by allowing me to *always* dispatch to both the "ordinary" and "special" sides, even if they might not actually be present.

- Renamed `TupleType` and `TupleVal` over to `TuplePseudoType` and `TuplePseudoval` to recognize the fact that we might actually need/want *real* tuples in the type system, to go along with these fake ones (that need to be optimized away).

The biggest doubt I have about this change is the whole `FilteredTupleType` thing; it seems like an obviously contrived type to add to the front-end type system that really only solves IR-level problems. A cleaner approach might have been to just add a plain old `TupleType` to the front-end type system (and initially I started with that), and then have yet another `LegalType`/`LegalVal` case that handles mapping from the fields of the original type to the numbered tuple elements.

I expect we'll actually want to make that change in the future (especially if we ever add true tuples to the front-end), but for right now I let myself be swayed by the desire to have these stripped/filtered types get names that explain their provenance ("where they came from") to make our output code more debuggable. The way I've done it is probably overkill, though, and we need a much more complete effort on the readability and debuggability of our output before anything like that is worth worrying about.

### CODE REVIEW

**Comment by tfoleyNV** [comment]
@csyonghe As I alluded to yesterday, the big design choice that I wanted to talk about here is how to represent the types that result when you take a mixed uniform/resource struct type and then strip/filter it to get a new aggregate type with just the uniform fields.

This change takes the approach of making a truly one-off type that represents the fact that we did this "stripping," but that can therefore retain things like field names (and hypothetically also re-use layout info) from the original type.

If we didn't care about what the output source code looked like, then the other clear option is just to have a real AST-type-system notion of a tuple type (distinct from the `LegalType::tuple` case which is a fake type meant to be eliminated) and use that to represent any stripped types. This option seems cleaner in the long term, but would require yet another `LegalType` case (to handle the mapping from field decl-refs over to tuple element indices), and then we'd have to figure out how much effort we could make on the post-IR emit logic to try to bring back usable type names.

I don't think I want the question to hold up this PR, but I wanted to make you aware of it so that you can give your input on what we want to do going forward.

**Comment by tfoleyNV** [comment]
Note: this change fixes #281.

Once I fix the build failure, at least. :)

**Comment by tfoleyNV** [comment]
The test failures here are happening because we are now emitting "temporary" names for constant buffers and global shader parameters that used to get friendly names because these tests were using the silly `-no-mangle` option.

I'm going to try to fix the failures, but this obviously points out that a lot of our current test cases are very brittle in the presence of the IR, because they depend on us producing output with certain textual names. This forces us to add flags that turn off the robust IR behavior (e.g., mangling names to avoid conflicts) and emit things that are more prone to accidental name clashes.

It seems like we need to expedite the shift from primary text-based output matching in our tests to something more robust. Either that *or* we need to use a more robust solution for checking test output (something that can weather changes to names).

File: source/slang/ir-legalize-types.cpp [hunk 1]
```diff
@@ -251,6 +440,442 @@ static bool isResourceType(Type* type)
...
+            pairInfo = new PairInfo();
+            pairInfo->elements = pairElements;
+        }
+
+        return LegalType::pair(ordinaryType, specialType, pairInfo);
+    }
+
+};
+
+static RefPtr<Type> createBuitlinGenericType(
```
Comment (csyonghe) [inline source/slang/ir-legalize-types.cpp]: A typo here?

File: tests/compute/cbuffer-legalize.slang.expected.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+3F800000
```
Comment (csyonghe) [inline tests/compute/cbuffer-legalize.slang.expected.txt]: Why isn’t the expected result 2 3 4 5?

File: source/slang/ir-legalize-types.cpp (hunk 1 above)
Comment (tfoleyNV) [reply source/slang/ir-legalize-types.cpp]: Good catch. Will fix this in the same commit where I try to fix the test failures.

File: tests/compute/cbuffer-legalize.slang.expected.txt (hunk 2 above)
Comment (tfoleyNV) [reply tests/compute/cbuffer-legalize.slang.expected.txt]: Hmm.. good point. I generated the expected output by first running the test case through `render-test` using our existing AST-to-AST path, so I relied on that code producing the right result.

I'm betting that something is going wrong and `render-test` is not setting up the constant buffer contents the way I'm expecting it to.

I'll debug it. Good catch again.

**Comment by csyonghe** [comment]
So my understanding of the difference between existing `FilteredTupleType` and the true `TupleType` is that `FilteredTupleType` includes addittional fields to track the origin of its elements, which is only intended for generating debuggable code? In that case why can't we just create a `TupleType` and add necessary decorations to it?

**Comment by tfoleyNV** [comment]
I think what you are describing is correct, just with the change that we can't attach decorations to the tuple type, but we *can* build a dictionary in the IR module that maps tuple types (or maybe other types more generally) to additional information that might inform emit logic.

There are two reason we can't attach things to the tuple type objects themselves:

1) Any two `TupleType` objects with the same elements are logically the "same" type, so just attaching a decoration to one of them won't suffice, because code later on my look at a different, but logically equivalent, type object and not see your decoration.

2) The decoration we'd like to attach to the type is specific to the particular target, entry point, etc. that we are outputting right now, and so we'd want to be careful not to leave cruft attached to types that might affect later output for other targets or entry points.

One last option I didn't talk about above, that would also be viable: we could actually have the type legalization pass construct its own dummy AST-level `struct` declarations to represent the stripped types, and then use those. That would conceivably let us do the pass without any new kinds of types making life complicated, but we'd still need some logic to handle field references so that we could map fields of the original type to the fields of the new type (and we'd similarly need to create new layouts that reflect the new fields).

File: tests/compute/cbuffer-legalize.slang.expected.txt (hunk 2 above)
Comment (tfoleyNV) [reply tests/compute/cbuffer-legalize.slang.expected.txt]: I've found at least one issue in `render-test`, but just fixing this one hasn't changed the test output. When I look at the debug log for the test I see:

```
D3D11 ERROR: ID3D11Device::CreateBuffer: Buffer with D3D11_BIND_CONSTANT_BUFFER cannot be combined with any other bind flags.
```

The code was trying to allocate the buffer with both the `D3D11_BIND_CONSTANT_BUFFER` and `...SHADER_RESOURCE` flags.

Comment (tfoleyNV) [reply tests/compute/cbuffer-legalize.slang.expected.txt]: And the other issue seems to be that there was no logic in place to actually bind the user-specified constant buffer(s).

**Pull request merged**